### PR TITLE
Make cache traceable

### PR DESCRIPTION
**Pull request by IlyasMoutawwakil**
# What does this PR do?

In #35792 I came to the conclusion that tensor subclassing is a process that can only be achieved currently with some restriction (e.g. [`get_seq_length()` needs to return a tensor](https://github.com/huggingface/transformers/pull/35792/files#diff-357aea2fcf1b51b2a8ad17c7d30044159b1fe1a3f6bd832cced70aca84c8be66R526)), adding more developer cognitive load when adding a cache class. In this PR we make the cache traceable and exportable by not being a `Module` and registering cache tensors as buffers directly in `TorchExportableModuleWithStaticCache`.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/integrations/executorch.py [hunk 1]
```diff
@@ -72,9 +69,13 @@ def __init__(self, model: PreTrainedModel):
             config=self.model.config,
             batch_size=self.model.generation_config.cache_config.batch_size,
             max_cache_len=self.model.generation_config.cache_config.max_cache_len,
-            dtype=self.model.dtype,
             device=self.model.generation_config.cache_config.device,
+            dtype=self.model.dtype,
         )
+        for i in range(len(self.static_cache.key_cache)):
+            self.register_buffer(f"key_cache_{i}", self.static_cache.key_cache[i], persistent=False)
```
Comment (guangy10) [inline src/transformers/integrations/executorch.py:76]: Curious why is non-persistent preferred in your opinion? Probably doesn't matter too much for inference as it will always start with filling the cache with prompt tokens even if they are persistent

Comment (IlyasMoutawwakil) [reply src/transformers/integrations/executorch.py:76]: non-persistent buffers are not saved with the model's `state_dict`, I think in the case of a big model with long sequence length static cache, the cache tensors will have nonnegligible memory footprint when exporting+saving the model

**Comment by IlyasMoutawwakil** [comment]
Tests in `optimum-executorch` are passing as well https://github.com/huggingface/optimum-executorch/pull/4

**Review by ArthurZucker — COMMENTED** [review]
SUper cool, IMO just missing what this enables (as in, documentation about how to use this now ) might be in optimum directly?

**Review by ArthurZucker — COMMENTED** [review]
Also I remember we had a need for nn.Module: copy. We need to be able to copy the cache object (or clone) for prefix re-usage

**Comment by ArthurZucker** [comment]
Can you check that https://github.com/huggingface/huggingface-llama-recipes/blob/main/performance_optimization/prompt_reuse.py still works

**Review by guangy10 — APPROVED** [review]
Thank you for validating the changes in Optimum https://github.com/huggingface/optimum-executorch/pull/4, especially considering the inconvenience caused by the disruptions due to the migration to a new repository

**Review by gante — APPROVED** [review]
LGTM, thank you for double-cheking the changes in many places @IlyasMoutawwakil 🤗

**Review by ArthurZucker — APPROVED** [review]
Let's go if this test ran locally! 🚀

**Comment by tugsbayasgalan** [comment]
Hey guys, i work on the torch.export team at PyTorch and i just wanted to verify that this change only works for Static kv cache right?

**Comment by IlyasMoutawwakil** [comment]
> Hey guys, i work on the torch.export team at PyTorch and i just wanted to verify that this change only works for Static kv cache right?

Hey ! not sure I understand, the change applies to `Cache` class which is inherited by all cache classes. And specifically for torch.export it removes the need to have `Cache` a subclass of `torch.nn.Module`.

**Comment by IlyasMoutawwakil** [comment]
just updated the branch, will merge tonight after tests pass to be sure everything is good.

**Comment by eljandoubi** [comment]
@IlyasMoutawwakil @ArthurZucker @gante @guangy10  detach Cahce from torch.nn.Module remove the .float() method from it which causes an error when calling  convert_to_fp32
image

**Comment by IlyasMoutawwakil** [comment]
self.float() should only be called if self is tensor or module, can you explain why it's called in this case ?

**Comment by gante** [comment]
I saw the same stack trace in another issue, but the user didn't share a script to reproduce it. @eljandoubi could you kindly share a script to reproduce the issue? 🙏

**Comment by eljandoubi** [comment]
@gante  I'm afraid that I can't but I can provide guidance.When fine-tuning PaliGemma 2 mix in mixed precision, the evaluation in the training loop is done in fp32 and so `convert_to_fp32` is called.

**Comment by IlyasMoutawwakil** [comment]
This sounds like it was a silent error, because calling `.float()` on a `Cache` instance (when it was a `nn.Module`), doesn't actually do anything, since `Cache.key_cache` / `Cache.value_cache` are just lists and not module parameters.

In the case of fine-tuning I believe you need to pass `use_cache=False` so that Cache is not returned (and thus converting it is not attempted). We can implement `.float()` method that actually does the conversion of `Cache.key_cache` / `Cache.value_cache` depending on whether we want to silence this issue or force user to set `use_cache=False` when training, wdyt @gante ?

**Comment by eljandoubi** [comment]
Thanks for enlightening me. I think set `use_cache=False` automatically in the `Trainer` like when using gradient checkpointing. Plus ignore objects that have no `.float()` method in `_is_fp16_bf16_tensor`.

**Comment by gante** [comment]
(see https://github.com/huggingface/transformers/pull/37044)
